h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java

```diff
@@ -212,6 +212,27 @@ public void dump(String format) {
       }
     }
   }
+  
+  private static XGBoostParameters.Backend findSuitableBackend(XGBoostParameters p) {
+    if ( p._backend == XGBoostParameters.Backend.auto || p._backend == XGBoostParameters.Backend.gpu ) {
+      if (H2O.getCloudSize() > 1) {
+        Log.info("GPU backend not supported in distributed mode. Using CPU backend.");
+        return XGBoostParameters.Backend.cpu;
+      } else if (! p.gpuIncompatibleParams().isEmpty()) {
+        Log.info("GPU backend not supported for the choice of parameters (" + p.gpuIncompatibleParams() + "). Using CPU backend.");
+        return XGBoostParameters.Backend.cpu;
+      } else if (XGBoost.hasGPU(H2O.CLOUD.members()[0], p._gpu_id)) {
+        Log.info("Using GPU backend (gpu_id: " + p._gpu_id + ").");
+        return XGBoostParameters.Backend.gpu;
+      } else {
+        Log.info("No GPU (gpu_id: " + p._gpu_id + ") found. Using CPU backend.");
+        return XGBoostParameters.Backend.gpu;
+      }
+    } else {
+      Log.info("Using CPU backend.");
+      return XGBoostParameters.Backend.cpu;
+    }
+  }
```

Comment:
I separated the part of the code that does the backend decision from the part that sets the params based on selected backend and booster